rust/repositories.bzl

```diff
@@ -42,98 +55,272 @@ filegroup(
+
+def produce_tool_suburl(tool_name, target_triple, version, iso_date = None):
+    """Produces a fully qualified Rust tool name for URL
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+    """
+
+    if iso_date:
+        return "{}/{}-{}-{}".format(iso_date, tool_name, version, target_triple)
+    else:
+        return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def produce_tool_path(tool_name, target_triple, version):
+    """Produces a qualified Rust tool name
+
+    Args:
+      tool_name: The name of the tool per static.rust-lang.org
+      target_triple: The rust-style target triple of the tool
+      version: The version of the tool among "nightly", "beta', or an exact version.
+    """
+
+    return "{}-{}-{}".format(tool_name, version, target_triple)
+
+def load_arbitrary_tool(ctx, tool_name, param_prefix, tool_subdirectory, version, iso_date, target_triple):
+    """Loads a Rust tool, downloads, and extracts into the common workspace.
+
+    Args:
+      ctx: A repository_ctx.
+      tool_name: The name of the given tool per the archive naming.
+      param_prefix: The name of the versioning param if the repository rule supports multiple tools.
+      tool_subdirectory: The subdirectory of the tool files (wo level below the root directory of
+                         the archive. The root directory of the archive is expected to match
+                         $TOOL_NAME-$VERSION-$TARGET_TRIPLE.
+      version: The version of the tool among "nightly", "beta', or an exact version.
+      iso_date: The date of the tool (or None, if the version is a specific version).
+      target_triple: The rust-style target triple of the tool
+    """
+
+    _check_version_valid(version, iso_date, param_prefix)
+
+    tool_suburl = produce_tool_suburl(tool_name, target_triple, version, iso_date)
+    url = "https://static.rust-lang.org/dist/{}.tar.gz".format(tool_suburl)
+
+    tool_path = produce_tool_path(tool_name, target_triple, version)
+    ctx.download_and_extract(
+        url,
+        output = "",
+        sha256 = FILE_KEY_TO_SHA.get(tool_suburl) or "",
+        type = "tar.gz",
+        stripPrefix = "{}/{}".format(tool_path, tool_subdirectory),
+    )
+
+def load_rust_compiler(ctx):
+    """Loads a rust compiler and yields corresponding BUILD for it
+
+    Args:
+      ctx: A repository_ctx.
+    Returns:
+      The BUILD file contents for this compiler and compiler library
+    """
+
+    target_triple = ctx.attr.exec_triple
+    load_arbitrary_tool(
+        ctx,
+        iso_date = ctx.attr.iso_date,
+        param_prefix = "rustc_",
+        target_triple = target_triple,
+        tool_name = "rustc",
+        tool_subdirectory = "rustc",
```

Comment:
So grabbing cargo and/or rustfmt from this tarball requires another function + subdirectory, but should cache to the same tarball?

Reply:
I see a couple of questions here:

---

"another function + subdirectory"
Yes, the followup CL will have one more of these functions per tool, but I'm planning to put those options into a separate "tool bundle" repository rule provided that an argument isn't made for including them into the compiler repository rule.

"cache to the same (...)"
This actually serves the opposite purpose, sort of. The subdirectory param names the second level dir that the contents of the "tool" lives in. (FWIW, the top level dir is just the thing i called the "tool_suburl").

The param is used for extracting the contents from the second level dir so that a composite directory similar to a `rustup` toolchain can be produced where all of the binaries live in the same bin/ dir, and all of the libraries live under lib/rustlib/ further under $TARGET_TRIPLE/ (**not** $HOST_TRIPLE, BTW). It's a minor detail, but the files here aren't ever repacked after this point -- the individual tars are all just dumped into this new remote workspace.

There is a particular motivation for doing this, by the way. My initial impl pulled the stdlibs down separately, but rustc uses paths local to the rustc executable to find the stdlib, so the final dir structure needs to match the structure produced by rustup.

Some examples of  the "second level dir":
rustfmt-(...).tar.gz contains `/rustfmt-nightly-(...)/rustfmt-preview/bin/rustfmt`

In this case, the tool_subdirectory is rustfmt-preview.

rust-std-nightly-($TARGET_TRIPLE).tar.gz contains `/rust-std-nightly-($TARGET_TRIPLE)/rust-std-($TARGET_TRIPLE)/lib/rustlib/($TARGET_TRIPLE)/*.rlib`

In this case, the tool_subdirectory is "rust-std-(...)".